File: sdk/basin-js/src/query.ts
/**
 * `.from(table)` — REST query builder for `/rest/v1/:table`.
 *
 * Route source (verified): crates/basin-rest/src/server.rs:243-249 —
 * `GET|POST|PATCH|DELETE /rest/v1/:table`.
 *
 * The filter grammar is Basin's PostgREST-*style* dialect, implemented in
 * crates/basin-rest/src/parser.rs (`parse_query` / `parse_op`):
 *
 * - `select=<cols,csv>` (identifiers only — no embedded resources/renames)
 * - `<col>=<op>.<value>` with ops `eq|neq|gt|gte|lt|lte|in|is`
 *   (`in.(a,b,c)` parenthesised; `is.null` / `is.notnull`)
 * - `order=<col>[.asc|.desc][,…]`, `limit=N`, `offset=N`
 * - `cursor=<token>` keyset pagination (id-based, v0.1), `stream=true` NDJSON
 *
 * It is NOT full PostgREST: no `or=`, no `not.`, no `like|ilike`, no embedded
 * `select` trees, no `Prefer: return=representation`. Filters AND together.
 *
 * Response shapes (crates/basin-rest/src/routes/data.rs):
 * - plain GET → JSON array of rows
 * - GET with `limit` or `cursor` → `{ rows, next_cursor }`
 * - large results / `?stream=true` → NDJSON, final line
 *   `{"_basin_next_cursor": "…"}` when paginating
 * - POST → 201, `{ ok, tag }` (or rows); PATCH/DELETE → `{ ok, tag }`
 * - DELETE may surface 501 `E_ENGINE_UNSUPPORTED`
 */

import type { ClientContext } from "./http.js";
import { requestJson, requestRaw } from "./http.js";
import type { ExecTag, Page } from "./types.js";

type Scalar = string | number | boolean | null;

function literal(v: Scalar): string {
  if (v === null) return "null";
  return String(v);
}

export type Row = Record<string, unknown>;

/** Normalized result of a GET: rows plus the cursor when paginating. */
export interface QueryResult<T> {
  rows: T[];
  nextCursor: string | null;
}

export class QueryBuilder<T extends Row = Row>
  implements PromiseLike<QueryResult<T>>
{
  #ctx: ClientContext;
  #table: string;
  #query: Array<[string, string]> = [];
  #paginated = false;

  constructor(ctx: ClientContext, table: string) {
    this.#ctx = ctx;
    this.#table = table;
  }

  /** `select=<cols>` projection; omit or pass `*` for all columns. */
  select(columns?: string | string[]): this {
    const cols = Array.isArray(columns) ? columns.join(",") : (columns ?? "*");
    this.#query.push(["select", cols]);
    return this;
  }

  eq(column: string, value: Scalar): this {
    this.#query.push([column, `eq.${literal(value)}`]);
    return this;
  }

  neq(column: string, value: Scalar): this {
    this.#query.push([column, `neq.${literal(value)}`]);
    return this;
  }

  gt(column: string, value: Scalar): this {
    this.#query.push([column, `gt.${literal(value)}`]);
    return this;
  }

  gte(column: string, value: Scalar): this {
    this.#query.push([column, `gte.${literal(value)}`]);
    return this;
  }

  lt(column: string, value: Scalar): this {
    this.#query.push([column, `lt.${literal(value)}`]);
    return this;
  }

  lte(column: string, value: Scalar): this {
    this.#query.push([column, `lte.${literal(value)}`]);
    return this;
  }

  /** `<col>=in.(a,b,c)` — parenthesised list per parser.rs `parse_in_list`. */
  in(column: string, values: Scalar[]): this {
    this.#query.push([column, `in.(${values.map(literal).join(",")})`]);
    return this;
  }

  /** `<col>=is.null` / `<col>=is.notnull`. */
  is(column: string, value: "null" | "notnull"): this {
    this.#query.push([column, `is.${value}`]);
    return this;
  }

  /** `order=<col>.asc|desc` (repeatable). */
  order(column: string, opts: { ascending?: boolean } = {}): this {
    const dir = opts.ascending === false ? "desc" : "asc";
    this.#query.push(["order", `${column}.${dir}`]);
    return this;
  }

  /** `limit=N`. Switches the GET response to `{ rows, next_cursor }`. */
  limit(n: number): this {
    this.#query.push(["limit", String(n)]);
    this.#paginated = true;
    return this;
  }

  offset(n: number): this {
    this.#query.push(["offset", String(n)]);
    return this;
  }

  /** Resume keyset pagination from a `next_cursor` token. */
  cursor(token: string): this {
    this.#query.push(["cursor", token]);
    this.#paginated = true;
    return this;
  }

  // -------------------------------------------------------------------------
  // Execution
  // -------------------------------------------------------------------------

  /** Execute as GET and normalize both response shapes. */
  async run(): Promise<QueryResult<T>> {
    const body = await requestJson<unknown>(
      this.#ctx,
      "GET",
      `/rest/v1/${this.#table}`,
      { query: this.#query },
    );
    return normalizeGet<T>(body);
  }

  /** Execute as GET and return rows only. */
  async rows(): Promise<T[]> {
    return (await this.run()).rows;
  }

  /**
   * Execute as a paginated GET (`{ rows, next_cursor }` shape). Identical to
   * `run()` but documents intent when iterating with `.cursor()`.
   */
  async page(): Promise<Page<T>> {
    const r = await this.run();
    return { rows: r.rows, next_cursor: r.nextCursor };
  }

  /**
   * Execute as GET with `?stream=true` (NDJSON). Yields rows; a trailing
   * `{"_basin_next_cursor": …}` line is captured, not yielded.
   */
  async *stream(): AsyncGenerator<T, string | null, void> {
    const query: Array<[string, string]> = [...this.#query, ["stream", "true"]];
    const res = await requestRaw(this.#ctx, "GET", `/rest/v1/${this.#table}`, {
      query,
    });
    const text = await res.text();
    let nextCursor: string | null = null;
    for (const line of text.split("\n")) {
      const trimmed = line.trim();
      if (trimmed.length === 0) continue;
      const parsed = JSON.parse(trimmed) as Record<string, unknown>;
      if (typeof parsed["_basin_next_cursor"] === "string") {
        nextCursor = parsed["_basin_next_cursor"];
        continue;
      }
      yield parsed as T;
    }
    return nextCursor;
  }

  /** `POST /rest/v1/:table` — insert one object or an array (201). */
  async insert(values: Row | Row[]): Promise<ExecTag | T[]> {
    return requestJson(this.#ctx, "POST", `/rest/v1/${this.#table}`, {
      body: values,
    });
  }

  /** `PATCH /rest/v1/:table?<filters>` — update rows matching the filters. */
  async update(values: Row): Promise<ExecTag | T[]> {
    return requestJson(this.#ctx, "PATCH", `/rest/v1/${this.#table}`, {
      query: this.#query,
      body: values,
    });
  }

  /**
   * `DELETE /rest/v1/:table?<filters>`. May throw `E_ENGINE_UNSUPPORTED`
   * (501) on engines without DELETE support — see data.rs `delete_table`.
   */
  async delete(): Promise<ExecTag | T[]> {
    return requestJson(this.#ctx, "DELETE", `/rest/v1/${this.#table}`, {
      query: this.#query,
    });
  }

  /** Awaiting the builder executes the GET (`await client.from('t').eq(…)`). */
  then<R1 = QueryResult<T>, R2 = never>(
    onfulfilled?: ((value: QueryResult<T>) => R1 | PromiseLike<R1>) | null,
    onrejected?: ((reason: unknown) => R2 | PromiseLike<R2>) | null,
  ): PromiseLike<R1 | R2> {
    return this.run().then(onfulfilled, onrejected);
  }
}

function normalizeGet<T>(body: unknown): QueryResult<T> {
  if (Array.isArray(body)) {
    return { rows: body as T[], nextCursor: null };
  }
  if (typeof body === "object" && body !== null && "rows" in body) {
    const wrapped = body as { rows: T[]; next_cursor: string | null };
    return { rows: wrapped.rows ?? [], nextCursor: wrapped.next_cursor ?? null };
  }
  // `{ok, tag}` empty-result shape (ExecResult::Empty) → no rows.
  return { rows: [], nextCursor: null };
}

File: sdk/basin-go/query.go
package basin

import (
	"bufio"
	"context"
	"encoding/json"
	"fmt"
	"iter"
	"strconv"
	"strings"
)

// QueryBuilder is a fluent builder for /rest/v1/:table queries.
//
// Route source (verified): crates/basin-rest/src/server.rs:243-249 —
// GET|POST|PATCH|DELETE /rest/v1/:table.
//
// Filter grammar (Basin's PostgREST-style dialect, parser.rs):
//   - select=<cols,csv>
//   - <col>=<op>.<value> with ops eq|neq|gt|gte|lt|lte|in|is
//     (in.(a,b,c) parenthesised; is.null / is.notnull)
//   - order=<col>[.asc|.desc][,...], limit=N, offset=N
//   - cursor=<token> keyset pagination
//
// Filters AND together. There is no OR, NOT, LIKE/ILIKE, or embedded resource
// selects — this is not full PostgREST.
//
// Response shapes (crates/basin-rest/src/routes/data.rs):
//   - plain GET → JSON array of rows
//   - GET with limit or cursor → { rows, next_cursor }
//   - POST → 201, { ok, tag }; PATCH/DELETE → { ok, tag }
//   - DELETE may surface 501 E_ENGINE_UNSUPPORTED
type QueryBuilder struct {
	t     *transport
	table string
	query []queryParam
}

func newQueryBuilder(t *transport, table string) *QueryBuilder {
	return &QueryBuilder{t: t, table: table}
}

// clone returns a shallow copy so filter methods are chain-safe.
func (q *QueryBuilder) clone() *QueryBuilder {
	params := make([]queryParam, len(q.query))
	copy(params, q.query)
	return &QueryBuilder{t: q.t, table: q.table, query: params}
}

func (q *QueryBuilder) push(key, val string) *QueryBuilder {
	out := q.clone()
	out.query = append(out.query, queryParam{key, val})
	return out
}

// formatValue converts a typed filter value to the string literal expected by
// Basin's PostgREST-style parser. Rules:
//
//   - string → verbatim (no quoting; the server parser handles the op.<value> split)
//   - bool   → "true" / "false"
//   - integer types (int, int8…int64, uint…uint64) → decimal
//   - float32/float64 → shortest decimal representation
//   - nil    → "null"  (use with Is for IS NULL checks)
//   - anything else → fmt.Sprint(v)
func formatValue(v any) string {
	if v == nil {
		return "null"
	}
	switch t := v.(type) {
	case string:
		return t
	case bool:
		if t {
			return "true"
		}
		return "false"
	case int:
		return strconv.FormatInt(int64(t), 10)
	case int8:
		return strconv.FormatInt(int64(t), 10)
	case int16:
		return strconv.FormatInt(int64(t), 10)
	case int32:
		return strconv.FormatInt(int64(t), 10)
	case int64:
		return strconv.FormatInt(t, 10)
	case uint:
		return strconv.FormatUint(uint64(t), 10)
	case uint8:
		return strconv.FormatUint(uint64(t), 10)
	case uint16:
		return strconv.FormatUint(uint64(t), 10)
	case uint32:
		return strconv.FormatUint(uint64(t), 10)
	case uint64:
		return strconv.FormatUint(t, 10)
	case float32:
		return strconv.FormatFloat(float64(t), 'f', -1, 32)
	case float64:
		return strconv.FormatFloat(t, 'f', -1, 64)
	default:
		return fmt.Sprint(v)
	}
}

// Select sets the columns projection. Accepts a CSV string or a slice.
// Omitting columns or passing "*" selects all columns.
func (q *QueryBuilder) Select(columns ...string) *QueryBuilder {
	var cols string
	if len(columns) == 0 {
		cols = "*"
	} else {
		cols = strings.Join(columns, ",")
	}
	return q.push("select", cols)
}

// Eq adds an equality filter: column=eq.value.
// value may be any scalar type: string, int, int64, float64, bool, or nil.
func (q *QueryBuilder) Eq(column string, value any) *QueryBuilder {
	return q.push(column, "eq."+formatValue(value))
}

// Neq adds a not-equal filter: column=neq.value.
// value may be any scalar type: string, int, int64, float64, bool, or nil.
func (q *QueryBuilder) Neq(column string, value any) *QueryBuilder {
	return q.push(column, "neq."+formatValue(value))
}

// Gt adds a greater-than filter: column=gt.value.
// value may be any scalar type: string, int, int64, float64, bool, or nil.
func (q *QueryBuilder) Gt(column string, value any) *QueryBuilder {
	return q.push(column, "gt."+formatValue(value))
}

// Gte adds a greater-than-or-equal filter: column=gte.value.
// value may be any scalar type: string, int, int64, float64, bool, or nil.
func (q *QueryBuilder) Gte(column string, value any) *QueryBuilder {
	return q.push(column, "gte."+formatValue(value))
}

// Lt adds a less-than filter: column=lt.value.
// value may be any scalar type: string, int, int64, float64, bool, or nil.
func (q *QueryBuilder) Lt(column string, value any) *QueryBuilder {
	return q.push(column, "lt."+formatValue(value))
}

// Lte adds a less-than-or-equal filter: column=lte.value.
// value may be any scalar type: string, int, int64, float64, bool, or nil.
func (q *QueryBuilder) Lte(column string, value any) *QueryBuilder {
	return q.push(column, "lte."+formatValue(value))
}

// In adds an IN filter: column=in.(a,b,c).
// Each element may be any scalar type: string, int, int64, float64, bool, or nil.
func (q *QueryBuilder) In(column string, values []any) *QueryBuilder {
	parts := make([]string, len(values))
	for i, v := range values {
		parts[i] = formatValue(v)
	}
	return q.push(column, "in.("+strings.Join(parts, ",")+")")
}

// Is adds an IS filter for null/notnull: column=is.null or column=is.notnull.
// value must be the literal string "null" or "notnull".
func (q *QueryBuilder) Is(column, value string) *QueryBuilder {
	return q.push(column, "is."+value)
}

// Order appends an ordering directive: order=col.asc|desc.
func (q *QueryBuilder) Order(column string, ascending bool) *QueryBuilder {
	dir := "asc"
	if !ascending {
		dir = "desc"
	}
	return q.push("order", column+"."+dir)
}

// Limit sets a row limit, switching the response shape to { rows, next_cursor }.
func (q *QueryBuilder) Limit(n int) *QueryBuilder {
	return q.push("limit", fmt.Sprintf("%d", n))
}

// Offset sets a row offset.
func (q *QueryBuilder) Offset(n int) *QueryBuilder {
	return q.push("offset", fmt.Sprintf("%d", n))
}

// Cursor resumes keyset pagination from a NextCursor token.
func (q *QueryBuilder) Cursor(token string) *QueryBuilder {
	return q.push("cursor", token)
}

// Run executes the query as GET and returns a QueryResult.
func (q *QueryBuilder) Run(ctx context.Context) (*QueryResult, error) {
	raw, _, err := q.t.doJSON(ctx, "GET", "/rest/v1/"+q.table, q.query, nil, true)
	if err != nil {
		return nil, err
	}
	return normalizeGetResponse(raw)
}

// Stream executes the query with ?stream=true and returns an iterator that
// yields decoded rows as they arrive from the NDJSON response body. The
// iterator is pull-based (range-over-func, Go 1.23+):
//
//	for row, err := range q.Stream(ctx) {
//	    if err != nil { /* handle */ break }
//	    fmt.Println(row)
//	}
//
// The response body is read line-by-line with bufio.Scanner; each non-empty
// line is JSON-decoded into a Row. The trailing pagination line
// {"_basin_next_cursor":"..."} is captured and not yielded as a row — retrieve
// it via (*StreamResult).NextCursor after the loop.
//
// The caller must consume or break the iterator to ensure the HTTP response
// body is closed. Cancelling ctx terminates the iterator with the
// context's error.
func (q *QueryBuilder) Stream(ctx context.Context) iter.Seq2[Row, error] {
	return func(yield func(Row, error) bool) {
		params := make([]queryParam, len(q.query)+1)
		copy(params, q.query)
		params[len(q.query)] = queryParam{"stream", "true"}

		resp, err := q.t.doStream(ctx, "/rest/v1/"+q.table, params)
		if err != nil {
			yield(nil, err)
			return
		}
		defer resp.Body.Close()

		scanner := bufio.NewScanner(resp.Body)
		for scanner.Scan() {
			line := strings.TrimSpace(scanner.Text())
			if line == "" {
				continue
			}
			var parsed map[string]any
			if err := json.Unmarshal([]byte(line), &parsed); err != nil {
				if !yield(nil, fmt.Errorf("basin: stream decode: %w", err)) {
					return
				}
				continue
			}
			// Skip the trailing pagination sentinel line; do not yield it.
			if _, ok := parsed["_basin_next_cursor"]; ok {
				continue
			}
			if !yield(Row(parsed), nil) {
				return
			}
		}
		if err := scanner.Err(); err != nil {
			yield(nil, fmt.Errorf("basin: stream read: %w", err))
		}
	}
}

// StreamResult is returned by StreamWithCursor; it exposes the next_cursor
// from the trailing NDJSON sentinel line so callers can page through results.
type StreamResult struct {
	// NextCursor is the opaque cursor from the trailing
	// {"_basin_next_cursor":"..."} line, or "" when there are no more pages.
	NextCursor string
}

// StreamWithCursor is like Stream but also captures the trailing
// {"_basin_next_cursor":"..."} pagination sentinel line. The returned
// *StreamResult is populated after the iterator is fully consumed.
//
//	res := &basin.StreamResult{}
//	for row, err := range q.StreamWithCursor(ctx, res) {
//	    if err != nil { break }
//	    handle(row)
//	}
//	fmt.Println(res.NextCursor)
func (q *QueryBuilder) StreamWithCursor(ctx context.Context, out *StreamResult) iter.Seq2[Row, error] {
	return func(yield func(Row, error) bool) {
		params := make([]queryParam, len(q.query)+1)
		copy(params, q.query)
		params[len(q.query)] = queryParam{"stream", "true"}

		resp, err := q.t.doStream(ctx, "/rest/v1/"+q.table, params)
		if err != nil {
			yield(nil, err)
			return
		}
		defer resp.Body.Close()

		scanner := bufio.NewScanner(resp.Body)
		for scanner.Scan() {
			line := strings.TrimSpace(scanner.Text())
			if line == "" {
				continue
			}
			var parsed map[string]any
			if err := json.Unmarshal([]byte(line), &parsed); err != nil {
				if !yield(nil, fmt.Errorf("basin: stream decode: %w", err)) {
					return
				}
				continue
			}
			// Capture the pagination sentinel; do not yield it as a row.
			if cursor, ok := parsed["_basin_next_cursor"]; ok {
				if s, ok := cursor.(string); ok && out != nil {
					out.NextCursor = s
				}
				continue
			}
			if !yield(Row(parsed), nil) {
				return
			}
		}
		if err := scanner.Err(); err != nil {
			yield(nil, fmt.Errorf("basin: stream read: %w", err))
		}
	}
}

// Insert inserts one row or a slice of rows (POST /rest/v1/:table → 201).
// Pass a map[string]any or []map[string]any.
func (q *QueryBuilder) Insert(ctx context.Context, values any) (*ExecResult, error) {
	raw, _, err := q.t.doJSON(ctx, "POST", "/rest/v1/"+q.table, nil, values, true)
	if err != nil {
		return nil, err
	}
	if raw == nil {
		return &ExecResult{OK: true}, nil
	}
	var result ExecResult
	if err := json.Unmarshal(raw, &result); err != nil {
		// The server sometimes returns the inserted rows; return generic OK.
		return &ExecResult{OK: true}, nil
	}
	return &result, nil
}

// Update patches rows matching the current filters (PATCH /rest/v1/:table).
func (q *QueryBuilder) Update(ctx context.Context, values map[string]any) (*ExecResult, error) {
	raw, _, err := q.t.doJSON(ctx, "PATCH", "/rest/v1/"+q.table, q.query, values, true)
	if err != nil {
		return nil, err
	}
	if raw == nil {
		return &ExecResult{OK: true}, nil
	}
	var result ExecResult
	if err := json.Unmarshal(raw, &result); err != nil {
		return &ExecResult{OK: true}, nil
	}
	return &result, nil
}

// Delete removes rows matching the current filters (DELETE /rest/v1/:table).
// May return *BasinError with Code "E_ENGINE_UNSUPPORTED" (501) on engines
// that do not support DELETE.
func (q *QueryBuilder) Delete(ctx context.Context) (*ExecResult, error) {
	raw, _, err := q.t.doJSON(ctx, "DELETE", "/rest/v1/"+q.table, q.query, nil, true)
	if err != nil {
		return nil, err
	}
	if raw == nil {
		return &ExecResult{OK: true}, nil
	}
	var result ExecResult
	if err := json.Unmarshal(raw, &result); err != nil {
		return &ExecResult{OK: true}, nil
	}
	return &result, nil
}

// normalizeGetResponse handles both plain array and { rows, next_cursor }
// response shapes from the Basin data endpoint.
func normalizeGetResponse(raw json.RawMessage) (*QueryResult, error) {
	if raw == nil {
		return &QueryResult{Rows: []Row{}}, nil
	}

	// Try array first.
	if len(raw) > 0 && raw[0] == '[' {
		var rows []Row
		if err := json.Unmarshal(raw, &rows); err != nil {
			return nil, fmt.Errorf("basin: decode rows: %w", err)
		}
		if rows == nil {
			rows = []Row{}
		}
		return &QueryResult{Rows: rows}, nil
	}

	// Try paginated object { rows, next_cursor }.
	var paged struct {
		Rows       []Row  `json:"rows"`
		NextCursor string `json:"next_cursor"`
	}
	if err := json.Unmarshal(raw, &paged); err != nil {
		return nil, fmt.Errorf("basin: decode paged response: %w", err)
	}
	if paged.Rows == nil {
		paged.Rows = []Row{}
	}
	return &QueryResult{Rows: paged.Rows, NextCursor: paged.NextCursor}, nil
}
